### ISSUE TITLE
The mavenjython plugin is breaking the maven build

### ISSUE THREAD

**Issue by raquel-ucl** [oracc/nammu#302]
The mavenjython plugin is an open source maven plugin we are using to compile the python modules that nammu needs into the final jar. 

I did some minor work to update it a while ago that is described in #35.

It's using `easy_install` and my plan from a while ago was to change the plugin so it uses `pip`, however I never had the time to do it, and it was low priority. Now it's breaking the build, so we need to fix this soon!

The last version of the plugin is 1.4 and you can see it [here](https://sourceforge.net/p/mavenjython/code/ci/master/tree/jython-compile-maven-plugin/src/main/java/net/sf/mavenjython/). You can clone the code and play with it locally:

```
git clone https://git.code.sf.net/p/mavenjython/code mavenjython-code
```

If I remember correctly, these are the steps needed to change and test the changes in the mavenjython plugin:

1. Edit `JythonMojo.java` and run `mvn clean compile package`. Note `package` is needed to create the jar in this case. The output jar will be in `jython-compile-maven-plugin/target/jython-compile-maven-plugin-2.0.jar`.
1. Place the jar in (or symlink from) your home maven folder, located in `~/.m2`. You'll need to create a `2.0` folder in `~/.m2/repository/net/sf/mavenjython/jython-compile-maven-plugin/`. Then, place the compiled jar and corresponding pom file/s, in the same way they are in `~/.m2/repository/net/sf/mavenjython/jython-compile-maven-plugin/1.4`. 
1. Recompile nammu, changing the mavenjython version in the pom file from 1.4 to 2.0. Maven will search for it in your `~/.m2` folder, and find it, otherwise it'll try to download it from the nexus maven repo and fail.
1. Once we get the 2.0 version, it'd be nice to put it on the nexus repository, but that takes time. I documented what I did in #35. 

This is mostly from memory so might not be 100% correct.

**Comment by raquel-ucl** [comment]
Anastasis and I have made some progress on this, but it's not fully fixed yet.

The plugin runs this command from within Nammu's `target/jython-plugins-tmp` folder (note you will only see this folder after running a mvn build in nammu): 

```
java -cp .:Lib -Dpython.home=. -Dpython.executable=jython org.python.util.jython Lib/site-packages/setuptools-0.6c11-py2.7.egg/easy_install.py --build-directory /Users/raquelalegre/workspace/ORACC/nammu/target/jython-plugins-tmp/build  <list of libraries to install>
```

We have changed the plugin so that it uses pip instead and runs this command:

```
java -cp .:Lib -Dpython.home=. -Dpython.executable=jython org.python.util.jython Lib/site-packages/bin/pip install --build-directory /Users/raquelalegre/workspace/ORACC/nammu/target/jython-plugins-tmp/build  <list of libraries to install>
```


For this, we packaged on a zip file all the files needed, and added a bin folder with the pip executable. These are the high level contents of the zip:

```
bin/
easy_install.py
easy_install.pyc
pip/
pip-9.0.1.dist-info/
pkg_resources/
setuptools/
setuptools-36.6.0.dist-info/
wheel/
wheel-0.30.0.dist-info/
```

The code is in  https://github.com/oracc/mavenjython-plugin.

So far, the new code for the plugin compiles and produces a new plugin jar. When using from nammu's build to prepare this list of libraries:

```
requests<2.12.0
pytest<3.0
pyyaml
mako
ply
jython-swingutils
https://github.com/ucl/pyoracc/archive/master.tar.gz
```

it breaks with this error message:

```
OSError: Cannot run program "/Users/raquelalegre/workspace/ORACC/nammu/target/jython-plugins-tmp/jython" (in directory "/Users/raquelalegre/workspace/ORACC/nammu/target/jython-plugins-tmp/build/pyyaml"): error=13, Permission denied
```

for pyyaml, mako, ply and pyoracc. If we comment those libraries out in the plugin list of libraries in Nammu's pom file, it works and the tests pass (which means at least it's working properly with pytests and requests).

Looking a bit more in detail and trying to isolate the problem just looking at mako's install by running:

```
mvn clean build
cd target/jython-plugins-tmp
java -cp .:Lib -Dpython.home=. -Dpython.executable=jython org.python.util.jython Lib/site-packages/bin/pip install --build-directory /Users/raquelalegre/workspace/ORACC/nammu/target/jython-plugins-tmp/build  mako
```

we find that the jython executable is actually 0B. In the plugin's code, the original author just has an empty fake jython file on purpose. However we have no idea what his purpose was because he didn't document it or comment the code!

I tried replacing the jython 0B file with a real one, and giving execute permission to all, and got this far:

```
Collecting mako
  Using cached Mako-1.0.7.tar.gz
    Complete output from command python setup.py egg_info:

    /Users/raquelalegre/workspace/ORACC/nammu/target contains neither jython-dev.jar nor jython.jar.
    Try running this script from the 'bin' directory of an installed Jython or
    setting $JYTHON_HOME.
    usage: jython [option] ... [-c cmd | -m mod | file | -] [arg] ...
    Try `jython -h' for more information.


    ----------------------------------------
Command "python setup.py egg_info" failed with error code 2 in /Users/raquelalegre/workspace/ORACC/nammu/target/jython-plugins-tmp/build/mako/
```

@ageorgou and I will keep on working on this today, since it's blocking all possible development on Nammu.

**Comment by raquel-ucl** [comment]
I forgot to add that we added the jython standalone jar `jython-standalone-2.7.1b3.jar` as `target/jython.jar`, rerun the java command and got:

```
  Using cached PyYAML-3.12.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    ImportError: No module named setuptools

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /Users/raquelalegre/workspace/ORACC/nammu/target/jython-plugins-tmp/build/pyyaml/
```

It seems the failing libraries all depend on setup tools. The ones that work have a wheel we can use to install.

**Comment by ageorgou** [comment]
Quick note: reverting to the old plugin version (1.4), it looks like the dependencies can be installed with easy_install by specifying ```--index-url="https://pypi.python.org/simple/"``` (this has been tested by running the executed ```java``` commands on the CLI individually, not by building the whole plugin). The exception is pyoracc, which gives a weird error (possibly from an underlying library):

```Downloading https://github.com/ucl/pyoracc/archive/master.tar.gz
error: [Errno -1] Unmapped exception: org.python.netty.util.IllegalReferenceCountException: refCnt: 0
```

and with `-vv`:

```
Downloading https://github.com/ucl/pyoracc/archive/master.tar.gz
Traceback (most recent call last):
  File "Lib/site-packages/setuptools-0.6c11-py2.7.egg/easy_install.py", line 5, in <module>
    main()
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 1712, in main
    with_ei_usage(lambda:
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 1700, in with_ei_usage
    return f()
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 1712, in <lambda>
    with_ei_usage(lambda:
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 211, in run
    self.easy_install(spec, not self.no_deps)
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 421, in easy_install
    download = self.package_index.download(spec, tmpdir)
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/package_index.py", line 392, in download
    found = self._download_url(scheme.group(1), spec, tmpdir)
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/package_index.py", line 614, in _download_url
    return self._attempt_download(url, filename)
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/package_index.py", line 621, in _attempt_download
    headers = self._download_to(url, filename)
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/package_index.py", line 568, in _download_to
    if fp: fp.close()
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/urllib.py", line 985, in close
    if self.fp: self.fp.close()
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/_socket.py", line 1960, in close
    self._sock.close()
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/httplib.py", line 566, in close
    fp.close()
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/_socket.py", line 1960, in close
    self._sock.close()
  File "/Users/ageorgou/nammu/target/jython-plugins-tmp/Lib/_socket.py", line 1078, in close
    msg.release()
	at org.python.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:101)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)

org.python.netty.util.IllegalReferenceCountException: org.python.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
```

**Comment by raquel-ucl** [comment]
Temporarily fixed so we can get by with easy_install and not changing the plugin. We can close this issue when #304 is merged, but we'll still need to asses #305 at some point.
